fix radius on input-group-label and input-group-button

andycochran opened the pull request:
This PR closes #8833 by explicitly setting styles for `.input-group-button` within `.input-group` so that it and `.input-group-label` have the expected border radius.

Commit:
fix radius on input-group-label and input-group-button

ncoden commented on scss/forms/_input-group.scss:
```diff
@@ -34,11 +34,21 @@ $input-prefix-padding: 1rem !default;
+        border-radius: if($global-text-direction == rtl, 0 $input-radius $input-radius 0, $input-radius 0 0 $input-radius);
+      }
+
+      &.input-group-button > * {
```
Selectors can be merged here are the code in the same. So:
```scss
&:not(.input-group-button), &.input-group-button > * {
   ...
}
```
(don't do that, see the comment above).

Otherwise we're just duplicating code.

andycochran replied:
I've DRY'd this. :)

ncoden commented on scss/forms/_input-group.scss:
```diff
@@ -34,11 +34,21 @@ $input-prefix-padding: 1rem !default;
 
     > :first-child {
-      border-radius: if($global-text-direction == rtl, 0 $input-radius $input-radius 0, $input-radius 0 0 $input-radius);
+      &:not(.input-group-button){
```
Foundation provide a SASS library, not only CSS components.
* We can't expect Input group button to be implemented as `.input-group-button` (user should be able to create custom components with input group mixins)
* Why only targetting Input group button ?

So I would rather go for:
```scss
&, &  > * { ... }
```

To be honnest, the `> *` is ugly and should be removed. But we'll keep it for backward compatibility. Adding a comment here like "don't do this" could be useful so it does not became an example.

andycochran replied:
Since `.input-group-button` is hardcoded later in the `foundation-form-prepostfix` mixin, can't we expect it to be implemented as `.input-group-button`?

ncoden replied:
Ah f**. I did not realized that Input Group was not _component-oriented_ at all !
Well, its still a problem: Why only targetting Input group button ?

This is generally a bad idea to make our CSS component having particular effects on each-others as the user should be able to use them with custom CSS alongside its own components. Input Group should have the same behavior no matter what is the class of the button we put it. This is the best way to make someone loose 1 hour: "Why when I put my own button there it is not rounded anymore."

ncoden reviewed (changes requested):
Three things to keep in mind when adding selectors:
* Selectors specificity must be the lowest possbile
* Parents should not apply properties automatically on their children
* Duplicated properties

Commit:
dry input-group scss

andycochran commented:
@ncoden I say we live with this not being component-oriented, and make sure not to repeat this mistake in v7. Are there any remaining changes you require?

ncoden commented:
> Why only targetting Input group buttons ?

andycochran commented:
Because the docs explicitly state to place buttons inside the `input-group-button` wrapper.

andycochran commented:
We could probably get rid of the `> *` too: 
`&.input-group-button .button `

ncoden commented:
We should assume as little as possible to HTML structure inside our components. People should be able to freely use the classes of components/elements we provide without limited by self-references and dependencies (like `.button` in a selector that have nothing to do with our Button components).

> Because the docs explicitly state to place buttons inside the input-group-button wrapper.

Got it. So I think we can drop `&:not(.input-group-button)` and use:
```scss
&, &.input-group-button { ... }
```
As `.input-group-button` does not have any visual properties.

ncoden commented:
@andycochran What do you think ?

ncoden commented:
@andycochran Hey 👋. Will you have some time work on this ? I'd like your opinon on https://github.com/zurb/foundation-sites/pull/11104#issuecomment-377585825

ncoden committed:
refactor: simplify Input Group border-radius selector

We should assume as little as possible to HTML structure inside our components. People should be able to freely use the classes of components/elements we provide without limited by self-references and dependencies.

This commit simplify the selector used to apply border-radius on Input Group childreens to make all childreens affected, with a single exception for `.input-group-button` where sub-childreens are affected too.

ncoden commented:
I applied the requested changes

> We should assume as little as possible to HTML structure inside our components. People should be able to freely use the classes of components/elements we provide without limited by self-references and dependencies.
> 
> This commit simplify the selector used to apply border-radius on Input Group childreens to make all childreens affected, with a single exception for `.input-group-button` where sub-childreens are affected too.

Looks ready for stable v6.5.0

DanielRuf commented:
Oh, node-sass failing on Node.js 10?

ncoden committed:
Merge branch 'develop' into zurb/andycochran/8833-input-group-label-radius
